File: 3160.py
# 
from typing import List

from typing import List

class Solution:
    def queryResults(self, limit: int, queries: List[List[int]]) -> List[int]:
        unique_colors = 0
        ball_color_map = {}  # Maps ball -> color
        color_count_map = {}  # Maps color -> count
        result = []

        for ball, color in queries:
            if ball not in ball_color_map:
                ball_color_map[ball] = color
                color_count_map[color] = color_count_map.get(color, 0) + 1
                if color_count_map[color] == 1:
                    unique_colors += 1
            else:
                prev_color = ball_color_map[ball]
                if prev_color != color:
                    color_count_map[prev_color] -= 1
                    if color_count_map[prev_color] == 0:
                        unique_colors -= 1

                    ball_color_map[ball] = color
                    color_count_map[color] = color_count_map.get(color, 0) + 1
                    if color_count_map[color] == 1:
                        unique_colors += 1

            result.append(unique_colors)

        return result

File: 3160.cpp
// 3160. Find the Number of Distinct Colors Among the Balls
#include <bits/stdc++.h>
using namespace std;

class Solution{
public:
    vector<int> queryResults(int limit, vector<vector<int>>& queries){
        int unique_colors = 0;
        map<int, int> ball_color_map;
        map<int, int> color_count_map;
        vector<int> result;

        for (auto i: queries){

            int ball = i[0], color = i[1];
            if (ball_color_map.find(ball) == ball_color_map.end()){
                // unique_colors += 1;
                ball_color_map[ball] = color;
                if (++color_count_map[color] == 1)
                    unique_colors += 1;
            } else {
                int prev_color = ball_color_map[ball];
                if (prev_color != color){
                    if (--color_count_map[prev_color] == 0)
                        unique_colors--;

                    ball_color_map[ball] = color;
                    if (++color_count_map[color] == 1)
                        unique_colors++;
                }
            }

            result.push_back(unique_colors);            
        }

        return result;
    }
};

int main(){
    return 0;

}